### PR TITLE
Adds LR decay scheduler.

### PR DESCRIPTION
**Pull request by Adamits**
- Adds schedulers.add_argparse_args
- Refactors _get_lr_scheduler

I added some other functions to schedulers, to support converting from yoyodyne argparse args, to the torch scheduler args. We may also want to consider exposing the interval and frequency of schedule updates to the CLI. In this way, the `total_decay_steps` for the decay scheduler could be total_decay_iters, and depend on whether you request that the scheduler potentially update every step or epoch.

Also -- when I run my local `black`, it seems to modify most files, so I reverted that and just used my flake8 config. Is there anywhere I can see the linting setup for this repo, and apply it to my local?

Addresses #28 28

### CODE REVIEW

**Comment by kylebgorman** [comment]
Thanks for this Adam.

Re: the `black` config stuff, the config it runs is [here](https://github.com/CUNY-CL/yoyodyne/blob/2e15a37e4666195fbcf33d75834a8d504e65d3d6/.circleci/config.yml#L51).

Re: this design, I wonder if it would be easier if you just wrapped the Torch thing and renamed things as you feel? That ought to get rid of at least `_map_scheduler_kwargs`?  Then we could make our scheduler constructors ignore `**kwargs` too (like `def __init__(arg_i_care_about: T, **kwargs): ...`), and we don't need `filter_scheduler_kwargs` either, I think. However it is extremely possible if I am missing something here.

I don't really understand how to use this either. I wonder if we can do more with clear help strings here. (Sadly the Torch source code is not very helpful.)

**Comment by Adamits** [comment]
> Re: this design, I wonder if it would be easier if you just wrapped the Torch thing and renamed things as you feel? That ought to get rid of at least _map_scheduler_kwargs? Then we could make our scheduler constructors ignore **kwargs too (like def __init__(arg_i_care_about: T, **kwargs): ...), and we don't need filter_scheduler_kwargs either, I think. However it is extremely possible if I am missing something here.

No I think this does sound better. I will do it this way (probably wont get to it until tmrw).

> I don't really understand how to use this either. I wonder if we can do more with clear help strings here. (Sadly the Torch source code is not very helpful.)

Is this wrt torch schedulers?

**Comment by kylebgorman** [comment]
> > Re: this design, I wonder if it would be easier if you just wrapped the Torch thing and renamed things as you feel? That ought to get rid of at least _map_scheduler_kwargs? Then we could make our scheduler constructors ignore **kwargs too (like def **init**(arg_i_care_about: T, **kwargs): ...), and we don't need filter_scheduler_kwargs either, I think. However it is extremely possible if I am missing something here.
> 
> No I think this does sound better. I will do it this way (probably wont get to it until tmrw).

Cool. I think this would just be straight inheritance---tweaking the `__init__` and that's it.

> Is this wrt torch schedulers?

Yeah. I wonder if we need some docs about how to use the two schedulers we've enabled. Maybe that's all that's necessary. I also was having trouble understanding the flag `help` but also the docs in the `torch.optim` module itself for this class,

File: yoyodyne/schedulers.py [hunk 1]
```diff
@@ -64,6 +71,41 @@ def lr_lambda(self, step: int) -> float:
         return self.decay_factor * step**-0.5
 
 
+class LinearDecay(optim.lr_scheduler.LinearLR):
+    """Linear decay scheduler."""
+
+    optimizer: optim.Optimizer
```
Comment (kylebgorman) [inline yoyodyne/schedulers.py:77]: I think these type decls are no longer needed since the superclass can take care of it.

File: yoyodyne/schedulers.py [hunk 2]
```diff
@@ -88,63 +130,28 @@ def add_argparse_args(parser: argparse.ArgumentParser) -> None:
         "--start_factor",
         type=float,
         default=1 / 3,
-        help="Start factor for lr decay. Default: %(default)s.",
+        help="The starting multiplier for the lr in LinearDecay. Default: %(default)s.",
```
Comment (kylebgorman) [inline yoyodyne/schedulers.py:133]: suggest "LR"; also on lines 138 and 145.

Above, for  warmup steps, it reads "(warmupinvsqrt scheduler only)". Here it reads "in LinearDecay". I don't care which style we use (yours may be better actually) but we should standardize.

**Review by kylebgorman — APPROVED** [review]
I think this is a big improvement. A few notes, mostly about docs and style, but otherwise LGTM.

**Comment by kylebgorman** [comment]
The internet says the issue with CircleCI will resolve if you log out of CircleCI and then log back in with GitHub credentials.

**Comment by Adamits** [comment]
> The internet says the issue with CircleCI will resolve if you log out of CircleCI and then log back in with GitHub credentials.

I think it wasn't even set up in my fork. Have now done so (after logging in via github).

**Comment by kylebgorman** [comment]
CircleCI stuff is working now (either because something I did or something you did, not sure). Can you fix these?

```
./yoyodyne/schedulers.py:6:1: F401 'inspect' imported but unused
./yoyodyne/schedulers.py:128:80: E501 line too long (102 > 79 characters)
./yoyodyne/schedulers.py:134:80: E501 line too long (117 > 79 characters)
```

May need to run `black` on that file too after the fact. Other than that LGTM again.

**Comment by Adamits** [comment]
This is weird.

```
(yoyodyne) (base) adamwiemerslage ~/nlp-projects/morphology/yoyodyne $ black --line-length=79 yoyodyne/schedulers.py 
All done! ✨ 🍰 ✨
1 file left unchanged.
```

**Comment by kylebgorman** [comment]
> This is weird.

It just doesn't break apart strings for you, as you see...wish it did.
